discovery_engine_core/core/src/stack/filters/semantic.rs

```diff
@@ -220,9 +222,17 @@ pub(crate) fn filter_semantically(
         Criterion::MaxClusters(max_clusters) => find_n_clusters(&dendrogram, max_clusters),
     };
 
+    // among documents with the same label, keep the one with heaviest source weight
     izip!(documents, labels)
-        .unique_by(|(_, label)| *label)
-        .map(|(document, _)| document)
+        .fold(HashMap::new(), |mut map, (doc, label)| {
```

Comment:
:art: even though the folding does the job, this could be improved by using `Itertools::into_grouping_map()`. the resulting `GroupingMap` also has some neat methods which may come in handy in the map operation below.

Reply:
thanks, I factored out a helper function `source_weight` and indeed the above simplifies nicely to
```rust
    izip!(labels, documents)
        .into_grouping_map()
        .max_by_key(|_label, doc| source_weight(doc, sources))
```